pkg/pillar/devicenetwork/arp.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2017-2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package devicenetwork
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+type arpPort struct {
+	ifName  string
+	macAddr string
+	addr    net.IP
+	subnet  net.IPNet
+}
+
+type arpEntry struct {
+	outIntf string
+	macAddr string
+	addr    net.IP
+	subnet  net.IPNet
+}
+
+func makeArpGroups(dns types.DeviceNetworkStatus) map[string][]arpPort {
+	arpGroups := map[string][]arpPort{}
+	for _, port := range dns.Ports {
+		// Get the IP address that belongs to given subnet
+		for _, addrInfo := range port.AddrInfoList {
+			subnet := &port.Subnet
+			if HostFamily(addrInfo.Addr) != syscall.AF_INET {
+				continue
+			}
+			if !subnet.Contains(addrInfo.Addr) {
+				continue
+			}
+			port := arpPort{
+				ifName:  port.IfName,
+				macAddr: port.MacAddr,
+				addr:    addrInfo.Addr,
+				subnet:  *subnet,
+			}
+			if group, ok := arpGroups[subnet.String()]; ok {
+				arpGroups[subnet.String()] = append(group, port)
+			} else {
+				arpGroups[subnet.String()] = []arpPort{port}
+			}
+			break
+		}
+	}
+	return arpGroups
+}
+
+func makeArpEntries(log *base.LogObject, arpGroups map[string][]arpPort) []arpEntry {
+	arpEntries := []arpEntry{}
+	for key, group := range arpGroups {
+		if len(group) <= 1 {
+			// No ARP entries to be programmed
+			continue
+		}
+
+		_, subnet, err := net.ParseCIDR(key)
+		if err != nil {
+			log.Errorf("makeArpEntries: Failed ParseCIDR for subnet %s", key)
+			continue
+		}
+		for i := 0; i < len(group); i++ {
+			from := group[i]
+			for j := i + 1; j < len(group); j++ {
+				to := group[j]
+				entry := makeArpEntry(from.ifName, to.macAddr, to.addr, *subnet)
+				arpEntries = append(arpEntries, entry)
+
+				// Create reverse entry at the same time
+				entry = makeArpEntry(to.ifName, from.macAddr, from.addr, *subnet)
+				arpEntries = append(arpEntries, entry)
+			}
+		}
+	}
+	return arpEntries
+}
+
+func (my arpEntry) Equal(your arpEntry) bool {
+	if my.outIntf != your.outIntf ||
+		my.macAddr != your.macAddr ||
+		my.subnet.String() != your.subnet.String() ||
+		my.addr.String() != your.addr.String() {
+		return false
+	}
+	return true
+}
+
+func makeArpEntry(outIntf string, macAddr string, addr net.IP, subnet net.IPNet) arpEntry {
+	return arpEntry{
+		outIntf: outIntf,
+		macAddr: macAddr,
+		addr:    addr,
+		subnet:  subnet,
+	}
+}
+
+func arpEntriesDifferent(oldEntries []arpEntry, newEntries []arpEntry) bool {
+	if len(oldEntries) != len(newEntries) {
+		return true
+	}
+
+	for i := 0; i < len(oldEntries); i++ {
+		oldEntry := oldEntries[i]
+		newEntry := newEntries[i]
+
+		if !oldEntry.Equal(newEntry) {
+			return true
+		}
+	}
+	return false
+}
+
+func printArpEntries(log *base.LogObject, entries []arpEntry) {
+	for _, e := range entries {
+		log.Errorf("ARP entry - Out interface: %s, mac: %s, addr: %s, subnet: %s",
```

Comment:
Why error? log.Noticef() should be sufficient if you want to make sure they appear by default.

Reply:
I missed changing from Errorf to Noticef before. Will do that.